Move Jolly and Depression into mood handling

coiax opened the pull request:
:cl: coiax
tweak: The Jolly and Depression quirks are now much more active. You can
expect the positive/negative moodlets to affect you around 50% of the
time.
/:cl:

- Previously, the probability of the Jolly/Depression moodlet firing was
  so low, you only had a less than 1/3 chance of seeing the 2 minute
  moodlet in an hour of play. Now they will be far more active.
  Depression's hardcore point value has increased, since it is far more
  negative now.

- The quirks now just add the TRAIT_JOLLY and TRAIT_DEPRESSION, while a
  seperate proc in the mood component handles the recurring moodlets (by
  using status effects for tracking the recurring nature).

- There is a new debug only TRAIT_NEUTRAL, for a recurring neutral mood
  that does nothing.

- Gonbolas now have a different positive mood message from before, but
  function in the same way.

coiax committed:
Move Jolly and Depression into mood handling

:cl: coiax
tweak: The Jolly and Depression quirks are now much more active. You can
expect the positive/negative moodlets to affect you around 50% of the
time.
/:cl:

- Previously, the probability of the Jolly/Depression moodlet firing was
  so low, you only had a less than 1/3 chance of seeing the 2 minute
  moodlet in an hour of play. Now they will be far more active.
  Depression's hardcore point value has increased, since it is far more
  negative now.

- The quirks now just add the TRAIT_JOLLY and TRAIT_DEPRESSION, while a
  seperate proc in the mood component handles the recurring moodlets (by
  using status effects for tracking the recurring nature).

- There is a new debug only TRAIT_NEUTRAL, for a recurring neutral mood
  that does nothing.

- Gonbolas now have a different positive mood message from before, but
  function in the same way.

coiax committed:
Merge branch 'master' into i-feel-more-neutral

coiax commented:
This is a rework of #55774, to move the "recurring" nature of the traits into status effects, and keeping the old TRAITs, and making it so other sources of them other than quirks will also fire those status effects.

Qustinnus commented:
The status effects seem redundant, and checking for a traits existence every process() is wasteful. Just register for the signal of the trait being added and then run the behavior of "add this moodlet between x and y minutes".  The status effects have no real place in this system as mood handles everything a status effect should do.

coiax committed:
Merge branch 'master' into i-feel-more-neutral

coiax committed:
Rework mood events and traits to be signal based

The recurring mood is now a proc that uses a timer to call itself, and
is started by an "add trait" signal listener.

Rohesie commented on code/__DEFINES/traits.dm:
```diff
@@ -429,6 +432,8 @@ Remember to update _globalvars/traits.dm if you're adding/removing/renaming trai
 #define PAI_FOLDED "pai-folded"
 /// Trait applied to brain mobs when they lack external aid for locomotion, such as being inside a mech.
+/// Trait granted by [/datum/status_effect/gonbola_pacify]
+#define GONBOLA_TRAIT "gonbola"
```
I'm not entirely sure why it's gonBola in the code, but it should be gonDola.

Rohesie commented:
Pending a processing mood system @Qustinnus plans to PR to this branch.

Commit:
adds timed mood events

Commit:
need to actually add world.time

coiax committed:
Merge pull request #7 from Qustinnus/i-feel-more-neutral

adds timed mood events

Rohesie commented on code/__DEFINES/traits.dm:
```diff
@@ -176,8 +176,11 @@ Remember to update _globalvars/traits.dm if you're adding/removing/renaming trai
 #define TRAIT_HOLY				"holy"
-#define TRAIT_DEPRESSION		"depression"
-#define TRAIT_JOLLY				"jolly"
+/* Traits which cause "recurring moods", such as occasionally getting
+* a positive or negative moodlet for JOLLY and DEPRESSION respectively. */
+#define TRAIT_DEPRESSION        "depression"
+#define TRAIT_JOLLY             "jolly"
+#define TRAIT_NEUTRAL           "neutral"
```
```suggestion

/* Traits which cause "recurring moods", such as occasionally getting
* a positive or negative moodlet for JOLLY and DEPRESSION respectively. */
#define TRAIT_DEPRESSION        "depression"
#define TRAIT_JOLLY             "jolly"
#define TRAIT_NEUTRAL           "neutral"

```

Rohesie commented on code/datums/status_effects/debuffs.dm:
```diff
@@ -635,16 +635,17 @@
-	REMOVE_TRAIT(owner, TRAIT_PACIFISM, "gonbolaPacify")
-	REMOVE_TRAIT(owner, TRAIT_MUTE, "gonbolaMute")
-	REMOVE_TRAIT(owner, TRAIT_JOLLY, "gonbolaJolly")
+	REMOVE_TRAIT(owner, TRAIT_PACIFISM, GONBOLA_TRAIT)
+	REMOVE_TRAIT(owner, TRAIT_MUTE, GONBOLA_TRAIT)
+	SEND_SIGNAL(owner, COMSIG_CLEAR_MOOD_EVENT, GONBOLA_TRAIT)
```
```suggestion
	REMOVE_TRAIT(owner, TRAIT_PACIFISM, GONBOLA_TRAIT)
	REMOVE_TRAIT(owner, TRAIT_MUTE, GONBOLA_TRAIT)
	SEND_SIGNAL(owner, COMSIG_CLEAR_MOOD_EVENT, GONBOLA_TRAIT)
	return ..()
```

Rohesie commented on code/datums/components/mood.dm:
```diff
@@ -11,6 +11,14 @@
 	var/insanity_effect = 0 //is the owner being punished for low mood? If so, how much?
 	var/atom/movable/screen/mood/screen_obj
+	///Lazy lists of callbacks that are currently running on a "timer" managed by mood. This ia a list of callback || Next run time. This is used by things such as depression.
+	var/list/timed_mood_events
```
`QDEL_LIST` this on destruction

coiax committed:
Apply suggestions from code review

Co-authored-by: Rohesie <rohesie@gmail.com>

coiax committed:
Merge branch 'master' into i-feel-more-neutral

coiax committed:
Merge branch 'i-feel-more-neutral' of github.com:coiax/tgstation into i-feel-more-neutral

coiax committed:
Add recommended cleanup in Destroy

Rohesie commented:
Tested locally and working?

Cyberboss commented on code/datums/traits/negative.dm:
```diff
@@ -110,17 +110,13 @@
-/datum/quirk/depression/on_process(delta_time)
-	if(DT_PROB(0.05, delta_time))
-		SEND_SIGNAL(quirk_holder, COMSIG_ADD_MOOD_EVENT, "depression_mild", /datum/mood_event/depression_mild)
+	hardcore_value = 2
```
Undocumented change? What does this do?

coiax replied:
>    Previously, the probability of the Jolly/Depression moodlet firing was
>   so low, you only had a less than 1/3 chance of seeing the 2 minute
>    moodlet in an hour of play. Now they will be far more active.
>    Depression's hardcore point value has increased, since it is far more
 >   negative now.

That would be the hardcore point value, how many hardcore points you get if you escape on the shuttle with this quirk as one of your "hardcore random negative quirks".

coiax committed:
Merge branch 'master' into i-feel-more-neutral

coiax committed:
Reworks mood timers to be trait based

Instead of a mapping between CALLBACKS and time to fire, we instead just
store the trait, and the next time it will trigger a mood event, with
some tweaks to logic to ensure you can't use gain/loss of the underlying
trait to bypass cooldowns.

- Adds the "jolly grey cap", which gives you TRAIT_JOLLY when it is on
  your head. Maybe it will make someone happy at Christmas some day.

coiax commented:
@Rohesie These "recurring mood events" seem so awkward, I'm starting to consider just going back to the old way, of just having a probability of them firing, and then tune that probability so there's roughly 15 fires in an hour (2 minute runtime, 15 fires, 30 minutes up, around 50% uptime).

I admit that almost entirely guts everything I've worked on, but I'm trying to work out whether the fact it is difficult implies that I am overthinking it.

Thoughts? Trying not to sunk cost fallacy up in this.

coiax committed:
Merge branch 'master' into i-feel-more-neutral

Rohesie commented:
> 
> 
> @Rohesie These "recurring mood events" seem so awkward, I'm starting to consider just going back to the old way, of just having a probability of them firing, and then tune that probability so there's roughly 15 fires in an hour (2 minute runtime, 15 fires, 30 minutes up, around 50% uptime).
> 
> I admit that almost entirely guts everything I've worked on, but I'm trying to work out whether the fact it is difficult implies that I am overthinking it.
> 
> Thoughts? Trying not to sunk cost fallacy up in this.

Sometimes keeping things simple is better, but let's not over-engineer this.
Up to you at the end of the day, really.

coiax commented:
Superceded by #56580
